src/NServiceBus.Core/DelayedDelivery/TimeoutManager/TimeoutRecoverabilityBehavior.cs

```diff
@@ -0,0 +1,115 @@
+    using System.Threading.Tasks;
+
+    class TimeoutRecoverabilityBehavior : Behavior<ITransportReceiveContext>
+    {
+        public TimeoutRecoverabilityBehavior(string errorQueueAddress, string localAddress, IDispatchMessages dispatcher, BusNotifications notifications, CriticalError criticalError)
+        {
+            this.localAddress = localAddress;
+            this.errorQueueAddress = errorQueueAddress;
+            this.dispatcher = dispatcher;
+            this.notifications = notifications;
+            this.criticalError = criticalError;
+        }
+
+        public override async Task Invoke(ITransportReceiveContext context, Func<Task> next)
+        {
+            var message = context.Message;
+            var failureInfo = failures.GetFailureInfoForMessage(message.MessageId);
+
+            if (ShouldAttemptAnotherRetry(failureInfo))
+            {
+                try
+                {
+                    await next().ConfigureAwait(false);
+                    return;
+                }
+                catch (Exception exception)
+                {
+                    HandleProcessingFailure(context, message, exception, failureInfo);
+
+                    context.AbortReceiveOperation();
+                    return;
+                }
+            }
+
+            GiveUpForMessage(message, failureInfo);
+
+            await MoveToErrorQueue(context, message, failureInfo);
+        }
+
+        bool ShouldAttemptAnotherRetry(ProcessingFailureInfo failureInfo)
+        {
+            if (failureInfo == null)
+            {
+                return true;
+            }
+            return failureInfo.NumberOfFailedAttempts <= MaxNumberOfFailedRetries;
+        }
+
+        void HandleProcessingFailure(ITransportReceiveContext context, IncomingMessage message, Exception exception, ProcessingFailureInfo failureInfo)
+        {
+            failures.RecordFailureInfoForMessage(message.MessageId, exception);
+
+            Logger.Info($"First Level Retry is going to retry message '{message.MessageId}' because of an exception:", exception);
+
+            var numberOfFailedAttempts = failureInfo?.NumberOfFailedAttempts ?? 0;
+            notifications.Errors.InvokeMessageHasFailedAFirstLevelRetryAttempt(numberOfFailedAttempts, context.Message, exception);
+        }
+
+        void GiveUpForMessage(IncomingMessage message, ProcessingFailureInfo failureInfo)
+        {
+            message.Headers[Headers.FLRetries] = failureInfo.NumberOfFailedAttempts.ToString();
+
+            notifications.Errors.InvokeMessageHasFailedAFirstLevelRetryAttempt(failureInfo.NumberOfFailedAttempts, message, failureInfo.Exception);
+
+            failures.ClearFailureInfoForMessage(message.MessageId);
+
+            Logger.Info($"Giving up First Level Retries for message '{message.MessageId}'.");
+        }
+
+        async Task MoveToErrorQueue(ITransportReceiveContext context, IncomingMessage message, ProcessingFailureInfo failureInfo)
```

Comment:
@MarcinHoppe I think that the trade-off here is between readability and stability of algorithm in `Invoke`. From my personal experience I would not expect to have this kind of abstraction (`Invoke` is driving signatures of `MoveToErrorQueue` and `ShouldAttemptAnotherRetry` etc.) inside a single class. Another argument for readability is that passing only what methods needs enables the reader to have a better understanding of what method does. For example `ShouldAttemptAnotherRetry` if given integer presenting retry count gives me a better idea of what is happening than passing a whole class which includes also exception. Or with `HandleProcessingFailure` it's confusing why we are passing both `Exception` and `ProcessingFailureInfo`.